### PR TITLE
add separate host pre resource

### PR DESCRIPTION
**Pull request by xh3b4sd**
We want to extend the CF stack management. A first step is to separate the management a bit. Here is a new resource suggested which manages what we called "host pre stack" so far. The naming here is not clear so feel free to suggest something. For now I used "Control Plane Initializer", because it is the control plane initialization of the tenant if that makes sense.

### CODE REVIEW

File: service/controller/v24/resource/cpi/create.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		}
+
+		_, err = r.hostClients.CloudFormation.DescribeStacks(i)
+		if IsNotExists(err) {
+			// fall through
+		} else if err != nil {
+			return microerror.Mask(err)
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "found the tenant cluster's control plane initializer CF stack already exists")
+
```
Comment (kopiczko) [inline service/controller/v24/resource/cpi/create.go]: I think `canceling resource` log is missing.

File: service/controller/v24/key/key.go [hunk 2]
```diff
@@ -96,6 +96,10 @@ const (
 	WorkerASGRef                 = "workerAutoScalingGroup"
 )
 
+const (
+	CapabilityNamesIAM = "CAPABILITY_NAMED_IAM"
```
Comment (kopiczko) [inline service/controller/v24/key/key.go]: I would strongly prefer to put that inside the resource. This is very CF specific.

File: service/controller/v24/resource/cpi/create.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		var newAdapter adapter.Adapter
+		{
+			c := adapter.Config{
+				CustomObject:   customObject,
+				GuestAccountID: cc.Status.Cluster.AWSAccount.ID,
+				Route53Enabled: r.route53Enabled,
+			}
+
+			newAdapter, err = adapter.NewHostPre(c)
```
Comment (kopiczko) [inline service/controller/v24/resource/cpi/create.go]: Why not `hostPreAdapter`?

**Comment by xh3b4sd** [comment]
I was not sure about the implementation and naming and thought a draft is more appropriate. Just because of my own uncertainty.

File: service/controller/v24/key/key.go (hunk 2 above)
Comment (xh3b4sd) [reply service/controller/v24/key/key.go]: We will then have it in 4 different resources in the future: pre, master, worker, post. This here is pre.

File: service/controller/v24/resource/cpi/create.go (hunk 3 above)
Comment (xh3b4sd) [reply service/controller/v24/resource/cpi/create.go]: No string reason. The type is `Adapter`.

**Comment by kopiczko** [comment]
> I was not sure about the implementation and naming and thought a draft is more appropriate. Just because of my own uncertainty.

Yeah I think `cpi` is very cryptic name. I would rather prefer a full form. We can change the resource name any time though. I think this is a very good step into splitting this shit.

File: service/controller/v24/key/key.go (hunk 2 above)
Comment (kopiczko) [reply service/controller/v24/key/key.go]: I think at some point we can make a generic resource like `cfstack`. Then the problem disappears. I don't mind having it repeated 4 times. This isn't information shared between resources.

File: service/controller/v24/resource/cpi/create.go (hunk 1 above)
Comment (xh3b4sd) [reply service/controller/v24/resource/cpi/create.go]: I added logs here and in `delete.go`.

File: service/controller/v24/key/key.go (hunk 2 above)
Comment (xh3b4sd) [reply service/controller/v24/key/key.go]: Moved to the resource.

**Comment by xh3b4sd** [comment]
Going with this. Let me know when I should re-adjust.

**Pull request merged**